---
title: "atac timecourse analysis"
author: "Roman E. Reggiardo"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: hide
editor_options:
  chunk_output_type: console
---
# nb_setup
```{r, setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 10, collapse = T)
knitr::opts_chunk$set(message = F)
# here::set_here('/home/rreggiar/')
here::here()
suppressPackageStartupMessages(library(tidyverse))
```
# ggplot theme
```{r, include=FALSE}
library(ggplot2)
library(ggthemes)
base_size = 10
ggplot2::theme_set(ggthemes::theme_foundation(base_size = base_size,
                                              base_family = 'Helvetica') + 
                     theme(plot.title = element_text(size = base_size, face = 'bold'),
                           panel.background = element_rect(colour = NA),
                           plot.background = element_rect(colour = NA),
                           panel.border = element_rect(colour = NA), 
                           axis.line = element_line(), 
                           axis.line.x = NULL, 
                           axis.line.y = NULL, 
                           axis.text = element_text(size = rel(0.95)), 
                           axis.text.x = element_text(margin = 
                                                        margin(t = 0.8 * base_size/4)), 
                           axis.text.x.top = element_text(margin = 
                                                            margin(b = 0.8 * base_size/4), 
                                                          vjust = 0), 
                           axis.text.y = element_text(margin = 
                                                        margin(r = 0.5 * base_size/4), 
                                                      hjust = 1),
                           axis.text.y.right = element_text(margin = 
                                                              margin(l = 0.5 * base_size/4), 
                                                            hjust = 0), 
                           axis.ticks = element_line(), 
                           axis.ticks.length = unit(base_size/2.5, "pt"), axis.ticks.length.x = NULL, 
                           axis.ticks.length.x.top = NULL, axis.ticks.length.x.bottom = NULL, 
                           axis.ticks.length.y = NULL, axis.ticks.length.y.left = NULL, 
                           axis.ticks.length.y.right = NULL,
                           strip.text = element_text(size = rel(0.8), face = 'bold'),
                           strip.background = element_rect(size = 0.2),
                           legend.key.size= unit(0.08, "in"),
                           legend.spacing = unit(0, "in"),
                           legend.key = element_rect(colour = NA),
                           legend.title = element_text(face="italic"),
                           legend.text = element_text(face = 'bold'),
                           legend.justification = c("right", "top"),
                           legend.box.just = "right",
                           legend.margin = margin(6, 6, 6, 6),
                           plot.tag = element_text(size = base_size, face = 'bold'),
                           plot.margin=margin(0.04,
                                              0.04,
                                              0.04,
                                              0.04,
                                              unit = "in"),
                           legend.box.spacing = unit(-0.02, 'in'),
                           panel.grid.major = element_blank(),
                           panel.grid.minor = element_blank()
                           )
                   )
txt.mm_to_pts <- function(ratio){ base_size * ratio * (25.4 / 72.27) }
```

# Data
## load and parse
```{r, include=FALSE}
# these are hard-coded since the data files are all the same
peak_colnames <- c('peak_id', 'chr', 'start', 'end',
                   'strand', 'score', 'focus', 'anno',
                   'detail', 'distance_tss', 'nearest_prom',
                   'entrez_id', 'nearest_unigene', 'nearest_refseq',
                   'nearest_ensg', 'gene_name', 'gene_alias', 
                   'gene_desc', 'biotype')
# grab all the output directories
data_dirs <- list.dirs(here::here('data'), recursive = F)
# aggregate the output data w/in dirs to a list of DFs
data_agg <- 
  lapply(data_dirs, function(dir) { 
  
    dir_name <- basename(dir)
    
    lfc4_colnames <- c('peak_id', 'chr', 'start', 'end',
                     'strand', 'score', 'focus', 
                     'tot_tags', 'bg_tags', 'foldChange', 'pval')
    
    peak <- 
      vroom::vroom(
        list.files(dir, 
                   pattern = '.peak',
                   full.names = T), col_names = peak_colnames
      )
    
    lfc4 <- 
    vroom::vroom(
      list.files(dir, 
                 pattern = 'LFC4',
                 full.names = T), col_names = lfc4_colnames
      )
    
    # add identifying column and calculate log2FC
    lfc4$foldChange <- as.numeric(lfc4$foldChange)
    lfc4$comparison <- dir_name
    lfc4$log2FoldChange <- log2(lfc4$foldChange)
    

    peak_lfc4 <- merge(peak, lfc4, by = c('peak_id', 'chr', 
                                            'start', 'end', 
                                            'strand'))
    return(peak_lfc4)

   }
)
# remove NULL entries (AvA comparisons)
data_agg_filt <- data_agg[!sapply(data_agg, is.null)]
# bind results into single data frame
atac_data <- do.call('rbind', data_agg_filt)

```

# Feature engineering
## parse out useful variables
```{r, include=FALSE}
# make some columns to enable numeric comparisons of time, categorical of context
atac_tidy <- 
  atac_data %>% 
    mutate(enriched_in = str_split_fixed(comparison, pattern = '.vs.', n = 3)[,1],
           compared_to = str_split_fixed(comparison, pattern = '.vs.', n = 3)[,2]) %>% 
    mutate(enriched_time = gsub(str_split_fixed(enriched_in, pattern = '_', n = 2)[,2], 
                                pattern = 'hr', replacement = ''),
           compared_time = gsub(str_split_fixed(compared_to, pattern = '_', n = 2)[,2], 
                                pattern = 'hr', replacement = '')) %>% 
    mutate(enriched_context = str_split_fixed(enriched_in, pattern = '_', n = 2)[,1],
           compared_context = str_split_fixed(compared_to, pattern = '_', n = 2)[,1])
```


# Data viz
This is a preliminary analysis that has a lot more room for growth and potential both content-wise and in terms of aesthetics. More of a survey of the major events and trends I can find quickly -- I have been really swamped lately and have more time now and in the coming weeks to work on the data. In all the analysis below, I've removed the 24hr time points. I'll keep digging through this next week, let me know if you have any requests, questions, or ideas. Code is hidden for ease of viewing.  

## biotype time course
Over time, how many significantly enriched peaks are detected in each comparison, at each time point compared to T==0, colored by biotype.  

### number of significantly enriched peaks by biotype
```{r, echo=FALSE, warning=FALSE}
atac_tidy %>% 
  group_by(enriched_time, enriched_context, compared_context, biotype) %>% 
  filter(compared_time == '0',
         enriched_time != '24') %>% 
  summarize(count = n()) %>%
  mutate(comparison = paste0(enriched_context, ' vs ', compared_context)) %>% 
  ggplot(aes(enriched_time, count, color = biotype, group = biotype)) + 
  geom_line(alpha = 0.4) +
  geom_point(size = 2) +
  ylab('# sig. enriched peaks') + xlab('time') +
  facet_wrap(~comparison, ncol = 2, scales = 'free_x') +
  scale_color_brewer(palette = 'Dark2')
```

### fraction of significantly enriched peaks from each biotype
```{r, echo=FALSE, warning=FALSE}
atac_tidy %>% 
  group_by(enriched_time, enriched_context, compared_context, biotype) %>% 
  filter(compared_time == '0',
         enriched_time != '24') %>% 
  summarize(count = n()) %>%
  group_by(enriched_time, enriched_context, compared_context) %>% 
  mutate(frac = count/sum(count)) %>% 
  mutate(comparison = paste0(enriched_context, ' vs ', compared_context)) %>% 
  ggplot(aes(enriched_time, frac, color = biotype, group = biotype)) + 
  geom_line(alpha = 0.4) +
  geom_point(size = 2) +
  ylab('fraction of sig. enriched peaks') + xlab('time') +
  facet_wrap(~comparison, ncol = 2, scales = 'free_x') +
  scale_color_brewer(palette = 'Dark2')
```

## annotation time course
Same as above but instead of biotypes we're looking at the genomic annotations.  

### number of significantly enriched peaks by annotation
```{r, echo=FALSE}
atac_tidy %>% 
  mutate(anno = str_split_fixed(anno, ' \\(', n = 2)[,1]) %>% 
  group_by(enriched_time, enriched_context, compared_context, anno) %>% 
  filter(compared_time == '0',
         enriched_time != '24') %>% 
  summarize(count = n()) %>% 
  mutate(comparison = paste0(enriched_context, ' vs ', compared_context)) %>% 
  ggplot(aes(enriched_time, count, color = anno, group = anno)) + 
  geom_line(alpha = 0.4) +
  geom_point(size = 2) +
  ylab('# sig. enriched peaks') + xlab('time') +
  facet_wrap(~comparison, ncol = 2, scales = 'free_x') + 
  scale_color_brewer(palette = 'Dark2')
```

### fraction of significantly enriched peaks from each annotation
```{r, echo=FALSE}
atac_tidy %>% 
  mutate(anno = str_split_fixed(anno, ' \\(', n = 2)[,1]) %>% 
  group_by(enriched_time, enriched_context, compared_context, anno) %>% 
  filter(compared_time == '0',
         enriched_time != '24') %>% 
  summarize(count = n()) %>% 
  group_by(enriched_time, enriched_context, compared_context) %>% 
  mutate(frac = count/sum(count)) %>% 
  mutate(comparison = paste0(enriched_context, ' vs ', compared_context)) %>% 
  ggplot(aes(enriched_time, frac, color = anno, group = anno)) + 
  geom_line(alpha = 0.4) +
  geom_point(size = 2) +
  ylab('fraction of sig. enriched peaks') + xlab('time') +
  facet_wrap(~comparison, ncol = 2, scales = 'free_x') + 
  scale_color_brewer(palette = 'Dark2')
```

## individual peaks over time
Here I'm taking the peaks (and labeling with their associated gene) that only show up in all 4 time points for each comparison and plotting their log2Foldchange as a function of time. I think this can be expanded to characterize these 20 peaks/genes as consistent, burst, or linear/logistic responders.  

```{r, echo=FALSE}
atac_tidy %>% 
  filter(compared_time == '0',
         enriched_time != '24') %>% 
  mutate(comparison = paste0(enriched_context, ' vs ', compared_context)) %>% 
  dplyr::group_by(peak_id, comparison) %>%
  dplyr::summarise(n = dplyr::n(), .groups = "drop") %>%
  dplyr::filter(n == 4L) -> consistent_peak_list 

atac_tidy %>% 
  mutate(anno = str_split_fixed(anno, ' \\(', n = 2)[,1]) %>% 
  filter(compared_time == '0',
         enriched_time != '24') %>% 
  mutate(comparison = paste0(enriched_context, ' vs ', compared_context)) %>% 
  merge(consistent_peak_list, by = c('peak_id', 'comparison')) %>% 
  ggplot(aes(enriched_time, log2FoldChange, group = peak_id, color = anno)) +
  geom_line(alpha = 0.4) + 
  geom_point(size = 2) + 
  facet_wrap(~gene_name+comparison, scales = 'free_x') + 
  scale_color_brewer(palette = 'Dark2') +
  theme(legend.position = 'bottom')
```

# Clustering with ATAC signal
Going to try normalizing the alignments and using the resulting cpm at called peaks to cluster, should provide more signal than the diff accessibility data

## consensus peak list
```{r, include=FALSE}

### doesn't work because we end up with ~~no overlap/consensus
atac_tidy %>% 
  filter(compared_time == '0',
         enriched_time != '24') %>% 
  filter(enriched_context == 'Hi', compared_context == 'Hi') %>% 
  mutate(comparison = paste0(enriched_context, ' vs ', compared_context, '_', enriched_time)) %>% 
  select(peak_id, log2FoldChange, comparison) %>% 
  pivot_wider(values_from = log2FoldChange, names_from = comparison) %>% 
  drop_na()

  
```

## called peaks